snapshot/pkg/volume/gluster/snapshot.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gluster
+
+import (
+	"fmt"
+	"os/exec"
+
+	"github.com/golang/glog"
+	crdv1 "github.com/kubernetes-incubator/external-storage/snapshot/pkg/apis/crd/v1"
+	"github.com/kubernetes-incubator/external-storage/snapshot/pkg/cloudprovider"
+	"github.com/kubernetes-incubator/external-storage/snapshot/pkg/volume"
+	"github.com/pborman/uuid"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	glusterfsBinary = "gluster"
+	glusterfsEp     = "glusterfs-cluster"
+)
+
+type glusterfsPlugin struct {
+}
+
+var _ volume.Plugin = &glusterfsPlugin{}
+
+// RegisterPlugin registers the volume plugin
+func RegisterPlugin() volume.Plugin {
+	return &glusterfsPlugin{}
+}
+
+// GetPluginName gets the name of the volume plugin
+func GetPluginName() string {
+	return "glusterfs"
+}
+
+func (h *glusterfsPlugin) Init(_ cloudprovider.Interface) {
+}
+
+func (h *glusterfsPlugin) SnapshotCreate(pv *v1.PersistentVolume, tags *map[string]string) (*crdv1.VolumeSnapshotDataSource, *[]crdv1.VolumeSnapshotCondition, error) {
+	spec := &pv.Spec
+	if spec == nil || spec.Glusterfs == nil {
+		return nil, nil, fmt.Errorf("invalid PV spec %v", spec)
+	}
+
+	volumePath := spec.Glusterfs.Path
+	snapshotName := volumePath + string(uuid.NewUUID())
+	cmd := exec.Command("sudo", glusterfsBinary, "snapshot", "create", snapshotName, volumePath)
```

Comment:
I don't think "sudo" is a right thing to do... The controller must run under an user who can use the glusterfs utilities and sudo should not be required. The controller should also work when run in container.

Reply:
@tsmetana sure, addressed the comment.